### PR TITLE
KEYCLOAK-2940 - draft - Backchannel Logout

### PR DESCRIPTION
**Pull request by DaSmoo**
Hello,

we tried to implement the back-channel logout like it's described in the opendID specification. We would like to get technical feedback.

Open Points from our side:

- Why should we deprecate the old logout endpoint? The current GET logout will stay as it is, right?
- Should we validate the iat claim in the logout token?
- Should social oidc identity providers also support back-channel logout, e. g. google? 
- AuthenticationManager.backchannelLogoutClientSession()
> In the past the backchannel logout was always trigger even if no RP party session existed. Now if we want to implement the concept by 100% we would need to get a valid return value from the call of the back-channel logout. At this point we would like to ask you guys for your feedback for the best implementation that would also fit to the solution. 
 
OpenId Backchannel logout Spec: https://openid.net/specs/openid-connect-backchannel-1_0.html
RedHat ticket: https://issues.redhat.com/browse/KEYCLOAK-2940?_sscc=t

### CODE REVIEW

**Comment by stianst** [comment]
Haven't looked at this, and hoping @pedroigor can do so soon.

With regards to "Why should we deprecate the old logout endpoint? The current GET logout will stay as it is, right?" - do you mean the RP initiated logout endpoint? There are improvements/changes we need to do there, but that is a separate issue to backchannel logout.

**Comment by DaSmoo** [comment]
> 
> 
> Haven't looked at this, and hoping @pedroigor can do so soon.
> 
> With regards to "Why should we deprecate the old logout endpoint? The current GET logout will stay as it is, right?" - do you mean the RP initiated logout endpoint? There are improvements/changes we need to do there, but that is a separate issue to backchannel logout.

Hey @stianst,

yes in the Ticket: https://issues.redhat.com/browse/KEYCLOAK-2940?_sscc=t there it is written down that we should mark the old logout endpoint as deprecated. We asked ourselves if we should deprecate the old logout endpoint in our contribution and integrate the existing @GET part into our new LogoutEndpoint or if we should leave it as it is so our contribution doesn't get to extensive.

Greetings David

**Comment by pedroigor** [comment]
@DaSmoo Reviewing this one today ...

File: core/src/main/java/org/keycloak/util/TokenUtil.java
Comment (pedroigor) [inline core/src/main/java/org/keycloak/util/TokenUtil.java:55]: I would just reuse `org.keycloak.representations.IDToken#NONCE`.

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/ClientEntity.java
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/models/jpa/entities/ClientEntity.java:133]: Instead of changing DDL, could we just persist the new data as client attributes? Similar to what we do with `org.keycloak.protocol.oidc.OIDCConfigAttributes#JWKS_URL`.

It should help to avoid all the migration-related code herein ...

File: server-spi-private/src/main/java/org/keycloak/events/EventType.java
Comment (pedroigor) [inline server-spi-private/src/main/java/org/keycloak/events/EventType.java:39]: Nitpick, but it helps to have a more concise set of changes and make review easier. Could you revert this change?

**Review by pedroigor — CHANGES_REQUESTED** [review]
Nice work, tks for this one. A few initial comments as a start so we can move forward depending on what you think about my comments. Basically:

* Do not change DDL but re-use client attributes to store logout metadata. See comment.
* A bit of nitpicking, so you could potentially revert changes that are basically changing formating.

Depending on whether or not you think my comments make sense, we should be able to reduce the size of the changeset.

**Comment by pedroigor** [comment]
> > Haven't looked at this, and hoping @pedroigor can do so soon.
> > With regards to "Why should we deprecate the old logout endpoint? The current GET logout will stay as it is, right?" - do you mean the RP initiated logout endpoint? There are improvements/changes we need to do there, but that is a separate issue to backchannel logout.
> 
> Hey @stianst,
> 
> yes in the Ticket: https://issues.redhat.com/browse/KEYCLOAK-2940?_sscc=t there it is written down that we should mark the old logout endpoint as deprecated. We asked ourselves if we should deprecate the old logout endpoint in our contribution and integrate the existing @get part into our new LogoutEndpoint or if we should leave it as it is so our contribution doesn't get to extensive.
> 
> Greetings David

I would also keep the current one and deal with it later.

File: core/src/main/java/org/keycloak/util/TokenUtil.java
Comment (DaSmoo) [reply core/src/main/java/org/keycloak/util/TokenUtil.java:55]: sure i will change that to the IDToken#NONCE

Comment (benjamin37) [reply core/src/main/java/org/keycloak/util/TokenUtil.java]: done

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/ClientEntity.java
Comment (benjamin37) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/ClientEntity.java]: Done, we also remove the database migration files. No need to do database changes any more.

File: server-spi-private/src/main/java/org/keycloak/events/EventType.java
Comment (benjamin37) [reply server-spi-private/src/main/java/org/keycloak/events/EventType.java]: Sure, we reverted all formatting changes we found. Hope it's fine like this.

**Comment by benjamin37** [comment]
Hi @pedroigor,

thanks for your feedback. We changed the PR according to your comments. We still have an issue with the cross dc tests running on Travis CI. Do you have an Idea why these tests are failing?

Kind regards
David & Benjamin

**Comment by pedroigor** [comment]
@benjamin37 Thanks for the latest changes. 

Cross-DC tests are unstable, so don't worry about it. But I'm going to check how the clustering ones (single dc) are doing.

Regarding squashing commits, whatever you prefer. I like to go over the files again though, it helps me to understand more about the changes as well as catching anything I missed before :)

**Comment by stianst** [comment]
> > Haven't looked at this, and hoping @pedroigor can do so soon.
> > With regards to "Why should we deprecate the old logout endpoint? The current GET logout will stay as it is, right?" - do you mean the RP initiated logout endpoint? There are improvements/changes we need to do there, but that is a separate issue to backchannel logout.
> 
> Hey @stianst,
> 
> yes in the Ticket: https://issues.redhat.com/browse/KEYCLOAK-2940?_sscc=t there it is written down that we should mark the old logout endpoint as deprecated. We asked ourselves if we should deprecate the old logout endpoint in our contribution and integrate the existing @get part into our new LogoutEndpoint or if we should leave it as it is so our contribution doesn't get to extensive.
> 
> Greetings David

You can ignore updating the logout endpoint as part of this, we'll handle that as a separate issue. Take a look at https://issues.redhat.com/browse/KEYCLOAK-11603 for more details

File: services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java:138]: Would be the case to have a specific attribute to configure logout signatures instead of re-using the same attribute for ID tokens?

For me, looks like what you did is OK where the client should already support any signature set for ID tokens already.

**Review by pedroigor — CHANGES_REQUESTED** [review]

**Comment by pedroigor** [comment]
@benjamin37 Only a few last comments.

Also, it would be nice if we could have a better response to a backchannel logout when logout was not successful for some clients. Like a JSON listing the clients that were not successful, or something like that. But I think this is an improvement and not a blocker.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: I think we could at least return a `504` to indicate that a client wasn't successful? Wdyt?

As per 2.8.  Back-Channel Logout Response.

**Review by pedroigor — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java
Comment (DaSmoo) [reply services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java:138]: We used the ID token implementation because in the specification it says that the logout token should be encrypted and validated similar to the id token. In particular 2.6 "Logout Token Validation" it describes the use of the encryption, used for the ID token during client registration. So we tried to reuse the current id token implementation parts instead of adding new ones.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java
Comment (DaSmoo) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java:263]: Sure, we will add the logout parts to the existing logout endpoint. This was the part where we didn't know how far the deprecation will go, that's why we create a new Endpoint class.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
Comment (DaSmoo) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: From the point of the specification we would return an an 501 (see 2.7 Back-Channel Logout Actions). The problem here is that if we only have one client session (the OPs for example) a back-channel logout wouldn't make sense. In the past this part of the code has been always triggert and if it failed no actions or returns has been performed. But now it would be nice to know if this session that gets triggert with a back-channel logout is really a session that needs to be logout out via the back-channel logout or if it's just the OP's session that will be logged out later in the logout process ( see the method "backchannelLogout(KeycloakSession session, RealmModel realm, UserSessionModel userSession, UriInfo uriInfo,
 ClientConnection connection, HttpHeaders headers, boolean logoutBroker, boolean offlineSession" Class: AuthenticationManager.java line 233). If we now trigger a back-channel logout on the ops session it would fail because no IDP is configured.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java:263]: Yes, we can always refactor later to deprecate and have a new endpoint,

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: I see. I meant `504` to indicate whether or not some of the downstream sessions were not invalidated. `501` seems to be related to a failure during when processing the logout request?

Could you elaborate a bit more about the failure if no IDP is configured, please ?

Comment (DaSmoo) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: yes, the idea from the openID specification is when an error happens in the back-channel logout happens, an 501 should be returned. I would agree that it would be better to give a more detailed feedback which session failed. 

The problem is that this "back-channel logout light" (lets call it like that) is always used when an logout has been triggert. 
Currently the protocol.backchannelLogout() gets triggert but no boolean or any response getting returned. To fulfil the openID back-channel logout specification, we would need to return an http status. Now with our change we saw in a test scenario that, when there is only one session (the OP itself) and the logout gets triggert through the @GET public logout() this line of code also gets triggert and results in a false positive test with an incorrect return. 
Because at the end the session gets removed by "session.sessions().removeUserSession(realm, userSession);" even if the back-channel logout failed. The return would be 5xx but the session had been successfully removed through the removeUserSession() method. 

Maybe the best way would be to find an attribute or something that indicates the session as a non back-channel logout session. Then it would be easy when the logout process loops over all client sessions to decide which sessions should also be logout with the back-channel process. You will find this loop in the method AuthenticationManager.backchannelLogoutAll() line 324.

But we wasn't able to find something like that. Maybe this is only an problem when the OP is an Keycloak. 
If the OP is a Keycloak it always have an back-channel logout URL and supports the back-channel logout. So the OP would try to logout it self through a back-channel logout request. Which also results in a 5xx or 4xx error.

Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: I think I get it, see if it makes sense:

* I guess you mean `org.keycloak.protocol.oidc.OIDCLoginProtocol#backchannelLogout`, which we could change to return some information that the logout endpoint could use to decide which status code to return. At the very end of the method call you have a call to `org.keycloak.services.managers.ResourceAdminManager#sendLogoutRequest` which could return the status code that we could then use to build a Map where the key is the clientId and the value the status code. With this information, we could decide which status code to return.

* When there is a single session (the OP itself) and the current `logout` endpoint is called, I won't care much. But for the new endpoint, the expected behavior I think would be to just remove the session without involving a new logout request.

But yeah, I do think that proper status code is important for this new endpoint (not really for the current one).

**Review by pedroigor — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
Comment (benjamin37) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: I'm not sure about returning a JSON with clientId as Key and HTTP Status of the downstream backchannel logout as value as response body of the backchannel logout REST call.

Given the scenario you have an OP Keycloak with multiple Clients connected which are called by downstream RPs. Someone sends a backchannel logout request to the OP Keycloak. This someone does typically not know which clients are configured in the OP Keycloak. OP sends backchannel logout requests to downstream clients, let's assume one returns 200 and one 501. The respose of OP Keycloak would look something like this:

Status: `504 Gateway Timeout`

Body:
```json
{
  "client_1": 200,
  "client_2": 501
}
```

* The first problem we see is that there can be multiple sessions in each downstream client that need to be logged out with one backchannel logout request. So one status code would not be enough. We could also add session id's and so on to the json to get rid of that problem.
* The second problem is that "someone" who called the backchannel logout at the Keycloak OP does normally not know about client ids and session ids of the Keycloak OP. So that's not the kind of information that should be passed as a response right?

Our suggestion is to just return a 504 in case of a downstream client failing to logout but leaving the response body empty. What do you think?

Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:429]: Yeah, let's keep 504 and no body.

Let's try to come up with something that solves the original idea as a follow-up. We can do other things here like returning a link for force/re-try logout to specific clients, etc.

**Comment by pedroigor** [comment]
@DaSmoo Need a rebase. Please, let me know when you have the updates we discussed.

**Comment by benjamin37** [comment]
Hi @pedroigor,

sorry for the delay. We'll provide the updates and the rebase tomorrow.

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
Comment (benjamin37) [reply services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: Done

**Comment by benjamin37** [comment]
Hi @pedroigor,

we updated the PR. We'll also provide tests for downstream logouts, hope I can finish these ones today (or at least this week).

**Comment by pedroigor** [comment]
@DaSmoo Np. Latest changes look good.

**Comment by benjamin37** [comment]
Hi @pedroigor,

we added the tests for downstream logouts, you can have an other look.

**Review by pedroigor — DISMISSED** [review]
@DaSmoo Thanks.

**Comment by pedroigor** [comment]
@DaSmoo Could you rebase. Just noticed, after approving, that there are conflicts.

**Comment by benjamin37** [comment]
@pedroigor sure, I'll rebase again

**Comment by pedroigor** [comment]
@benjamin37 Forgot about something ... Could you also add this https://github.com/pedroigor/keycloak/commit/24e31b91e6cf79bbc67d38098f0b42a091eb2b9c to your change set ?

I know it is not something you did, but we should fix that part so that we can release connection properly.

**Comment by benjamin37** [comment]
@pedroigor did the rebase and integrated the changes of commit https://github.com/pedroigor/keycloak/commit/24e31b91e6cf79bbc67d38098f0b42a091eb2b9c

**Comment by pedroigor** [comment]
@benjamin37 Tks. Once CI finishes I'll merge.

**Comment by benjamin37** [comment]
Hi @pedroigor,

just wanted to ask if there is anything else we should do that' blocking the merge?

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@benjamin37 Merged, thanks !

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/OAuthClient.java
Comment (CG-ANY) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/OAuthClient.java:726]: There is a typo here. It was meant to be `logoutToken` instead of `logoutTokon`.

**Pull request merged**
